### PR TITLE
Lesions analysis

### PR DESCRIPTION
**Pull request by frheault**
@mdesco @arnaudbore yet another major new features...

Not ready for test, but could you read the docstring and help of both scripts and tell me if it is clear and if it is something you could see using later/soon?

I made the script based on project I could use it, but maybe something nice could be easily added and I didn't think of it.
This is a first working version, once merge it is possible I will have to slightly modify the json file for tractometry-flow, but the core will stay the same, this will be done in time.

PS: For stability I have to say how many digit I want to use for lesion count, do you think 001-999 is too low, do you think it is possible to have more than 999 lesions, what about 9999? Should it be user input?

### CODE REVIEW

File: scripts/scil_analyse_lesions_load.py [hunk 1]
```diff
@@ -0,0 +1,177 @@
+#!/usr/bin/env python
+# -*- coding: utf-8 -*-
+
+"""
+This script will output informations about lesions load of a bundle.
```
Comment (mdesco) [inline scripts/scil_analyse_lesions_load.py:5]: about lesion load in bundle(s).

File: scripts/scil_analyse_lesions_load.py [hunk 2]
```diff
@@ -0,0 +1,177 @@
...
+Either using as streamlines, binary map or a bundle voxel labels map.
```
Comment (mdesco) [inline scripts/scil_analyse_lesions_load.py:6]: The input can either be streamlines, binary bundle map, or a bundle voxel label map.

File: scripts/scil_analyse_lesions_load.py [hunk 3]
```diff
@@ -0,0 +1,177 @@
...
+    p1.add_argument('--bundle_labels_map',
+                    help='Path of the bundle labels_map.')
+
+    p.add_argument('--min_lesions_vol', type=float, default=7,
+                   help='Minimum lesions volume in mm3 [%(default)s].')
+    p.add_argument('--out_atlas',
+                   help='Save the lesions as an atlas.')
+    p.add_argument('--out_lesions_stats', action='store_true',
+                   help='Save the lesions-wise & streamlines count.')
+
```
Comment (mdesco) [inline scripts/scil_analyse_lesions_load.py:53]: I would remove pluriel to all lesions -> lesion
idem for labels -> label

File: scripts/scil_compute_connectivity.py [hunk 4]
```diff
@@ -27,6 +28,13 @@
 pre-computed maps (LABEL1_LABEL2.nii.gz) following the same naming convention
 as the input directory. Each will generate a matrix. The average non-zeros
 value in the map will be reported in the matrices nodes.
+
+The parameters --lesions_load will compute 3 lesions related matrices:
+lesions_count.npy, lesions_vol.npy, lesions_sc.npy and put it inside of a
+specified folder. They represent the number of lesions, the total volume of
+lesions and the total of streamlines going through the lesions for  of each
+connection. Each connection can be seen as a 'bundle' and then something
+similar to scil_analyse_lesions_load.py is run for each 'bundle'.
```
Comment (mdesco) [inline scripts/scil_compute_connectivity.py:37]: lesions -> lesion

**Review by mdesco — COMMENTED** [review]
Minor English edits.

Structure of input/output and options look good!

File: scripts/scil_analyse_lesions_load.py [hunk 5]
```diff
@@ -0,0 +1,154 @@
...
+from scilpy.utils.filenames import split_name_with_nii
+from scilpy.utils.metrics_tools import compute_lesion_stats
+
+
+def _build_arg_parser():
+    p = argparse.ArgumentParser(description=__doc__,
+                                formatter_class=argparse.RawTextHelpFormatter)
+
+    p.add_argument('in_lesion',
+                   help='Binary mask of the lesion (.nii.gz).')
```
Comment (mdesco) [inline scripts/scil_analyse_lesions_load.py:37]: lesion(s)

File: scripts/scil_analyse_lesions_load.py [hunk 6]
```diff
@@ -0,0 +1,154 @@
...
+    p.add_argument('out_json',
+                   help='Output file (.json) for lesion information')
+    p1 = p.add_mutually_exclusive_group()
+    p1.add_argument('--bundle',
+                    help='Path of the bundle file (.trk).')
+    p1.add_argument('--bundle_mask',
+                    help='Path of the bundle binary mask.')
+    p1.add_argument('--bundle_labels_map',
+                    help='Path of the bundle labels_map.')
```
Comment (mdesco) [inline scripts/scil_analyse_lesions_load.py:46]: please add (.nii.gz) for the above two options

File: scripts/scil_analyse_lesions_load.py [hunk 7]
```diff
@@ -0,0 +1,154 @@
+#!/usr/bin/env python
+# -*- coding: utf-8 -*-
+
+"""
+This script will output informations about lesion load in bundle(s).
+The input can either be streamlines, binary bundle map, or a bundle voxel
+label map.
```
Comment (mdesco) [inline scripts/scil_analyse_lesions_load.py:7]: Can you say 1-2 sentences about the min_lesion_volume parameter and how it comes into the script/algorithm please?

File: scripts/scil_compute_connectivity.py [hunk 8]
```diff
@@ -27,6 +28,13 @@
 pre-computed maps (LABEL1_LABEL2.nii.gz) following the same naming convention
 as the input directory. Each will generate a matrix. The average non-zeros
 value in the map will be reported in the matrices nodes.
+
+The parameters --lesion_load will compute 3 lesion related matrices:
+lesion_count.npy, lesion_vol.npy, lesion_sc.npy and put it inside of a
+specified folder. They represent the number of lesion, the total volume of
+lesion and the total of streamlines going through the lesion for  of each
```
Comment (mdesco) [inline scripts/scil_compute_connectivity.py:35]: the total number of streamlines (streamline count, sc) going through 

should lesion be replaced by lesion(s) everywhere?

**Review by mdesco — COMMENTED** [review]
I have made another round of small edits. 
Also, concerning your question above, I think 9999 lesions would be the safest (999 should be fine but we never know...). 

Let me know when it is ready to test. I have some MS data to test.

File: scripts/scil_analyse_lesions_load.py [hunk 9]
```diff
@@ -0,0 +1,154 @@
...
+                    help='Path of the bundle binary mask.')
+    p1.add_argument('--bundle_labels_map',
+                    help='Path of the bundle labels_map.')
+
+    p.add_argument('--min_lesion_vol', type=float, default=7,
+                   help='Minimum lesion volume in mm3 [%(default)s].')
+    p.add_argument('--out_lesion_atlas', metavar='FILE',
+                   help='Save the lesion as an atlas.')
+    p.add_argument('--out_lesion_stats', metavar='FILE',
+                   help='Save the lesion-wise.')
```
Comment (mdesco) [inline scripts/scil_analyse_lesions_load.py:53]: Save the lesion-wise what? Incomplete sentence.

File: scripts/scil_analyse_lesions_load.py [hunk 10]
```diff
@@ -0,0 +1,154 @@
...
+                    help='Path of the bundle file (.trk).')
+    p1.add_argument('--bundle_mask',
+                    help='Path of the bundle binary mask.')
+    p1.add_argument('--bundle_labels_map',
+                    help='Path of the bundle labels_map.')
+
+    p.add_argument('--min_lesion_vol', type=float, default=7,
+                   help='Minimum lesion volume in mm3 [%(default)s].')
+    p.add_argument('--out_lesion_atlas', metavar='FILE',
+                   help='Save the lesion as an atlas.')
```
Comment (mdesco) [inline scripts/scil_analyse_lesions_load.py:51]: Not quite sure what this sentence means at first read. i.e. "lesion as an atlas" 
What does that mean?

File: scripts/scil_analyse_lesions_load.py [hunk 11]
```diff
@@ -0,0 +1,154 @@
...
+    p.add_argument('--out_lesion_stats', metavar='FILE',
+                   help='Save the lesion-wise.')
+    p.add_argument('--out_streamlines_stats', metavar='FILE',
+                   help='Save the lesion-wise streamlines count.')
```
Comment (mdesco) [inline scripts/scil_analyse_lesions_load.py:55]: streamlines -> streamline

File: scripts/scil_analyse_lesions_load.py (hunk 9 above)
Comment (mdesco) [reply scripts/scil_analyse_lesions_load.py:53]: For the 3 --out options, please provide extension of the FILE you are expecting

**Review by mdesco — COMMENTED** [review]
If I test your command line with your data: braindata/processedData/frheault/max_lesions_PR.zip
Command: 
scil_analyse_lesions_load.py lesion_mask.nii.gz AF_R.json --bundle bundles/AF_R_m_cleaned.trk --out_lesion_atlas out_lesion_atlas.nii.gz --out_streamlines_stats stats.txt --out_streamlines_stats count.txt -f

I get the following error:
Traceback (most recent call last):
  File "/Users/desm2239/my_env/scilpy_master/bin/scil_analyse_lesions_load.py", line 7, in <module>
    exec(compile(f.read(), __file__, 'exec'))
  File "/Users/desm2239/Research/Source/scilpy/scripts/scil_analyse_lesions_load.py", line 154, in <module>
    main()
  File "/Users/desm2239/Research/Source/scilpy/scripts/scil_analyse_lesions_load.py", line 137, in main
    for key in lesion_dict.keys():
UnboundLocalError: local variable 'lesion_dict' referenced before assignment

Files output created:
out_lesion_atlas.nii.gz
AF_R.json

File: scripts/scil_compute_connectivity.py (hunk 8 above)
Comment (frheault) [reply scripts/scil_compute_connectivity.py:35]: It was lesions before and you said to switch it to lesion.
I will put lesion(s) everywhere in the docstring but variable name will stay `lesion`.
Is that ok?

Comment (mdesco) [reply scripts/scil_compute_connectivity.py:35]: yes ok

**Comment by frheault** [comment]
@mdesco The error you got is fixed (and the rest of the comment too). Sorry for that.
At least now you can test it. I will wait for the next pass and then fix the flow.

**Comment by scil-admin** [comment]
Build Failed :boom: 
https://jenkins.scil.usherbrooke.ca/job/scilpy/job/PR-371/11/

**Comment by scil-admin** [comment]
Build passed ! Good Job :beers: !

**Comment by scil-admin** [comment]
Build passed ! Good Job :beers: !

**Comment by scil-admin** [comment]
Build passed ! Good Job :beers: !

**Comment by scil-admin** [comment]
Build passed ! Good Job :beers: !

**Comment by scil-admin** [comment]
Build passed ! Good Job :beers: !

**Pull request merged**